### ISSUE TITLE
at-rule-name-space/newline-after

### ISSUE THREAD

**Issue by robcolburn** [stylelint/stylelint#1096]
- [x] `at-rule-name-space-after: "always"|"never"|"always-single-line"`
- [x] `at-rule-name-newline-after: "always"|"always-multiline-line"`

------------

Feature Request: Add a rule to enforce
* Adding of a space between `media` and it's opening parenthesis.
* Adding a space after closing parenthesis and before opening brace. 

```css
@media (max-width: 300px) {
/**   ↑                  ↑
 * The space before and after parentheses */
```

Similar To:
* `media-query-parentheses-space-inside `
* `media-feature-*-space-*`

**Comment by jeddy3** [comment]
@robcolburn Thanks for the suggestion.

You can enforce the space before the opening brace using the [`block-opening-brace-space-before`](https://github.com/stylelint/stylelint/blob/master/src/rules/block-opening-brace-space-before/README.md) rule.

There isn't a rule to enforce a space after the at-rule name *yet*, but it's a great suggestion.

I think two rules are needed here as they're quite a few use case considerations:

- `at-rule-name-space-after: "always"|"never"|"always-single-line"`
- `at-rule-name-newline-after: "always"|"always-multiline-line"`

That should allow the following code styles:

```css
/* single line at-rule with parens */
@media (width <= 100px) {}

/* spaced multiline at-rule with parens */
@media (width <=100px) and
  (orientation: portrait) {}

/* newlined multiline at-rule with parens */
@media
  (width <=100px) and
  (orientation: portrait) {}

/* single line at-rule with quoted params */
@import "a.css";

/* no spaced at-rules */
@import"a.css";
@media(width <= 100px) {}
```

I couldn't find anything in the specs about the validity of those last two, but PostCSS [does parse them](https://astexplorer.net/#/2F8sr0jNJs). Anyone got any thoughts on this?

Would you be interested in championing this rule i.e. helping us build it? It [looks like](https://astexplorer.net/#/2F8sr0jNJs) PostCSS tracks the whitespace after the at-rule name in the `raw.afterName` property and so these rules should (hopefully) be less complicated than the existing `*-space-*` and `*-newline-*` rules. Let us know if you're interested in contributing and we'll point you in the right direction :)

In the meantime, anyone else got thoughts on these two new rules?

**Comment by robcolburn** [comment]
@jeddy3 I could take a swing at it sure. Though I'm a total newbie to the codebase. ;) I've listened to talks on parsers l (Babel/ESLint), but I tend to get lost after they start talking about tree manipulations.

Honestly, I glazed past the `at-rule-*`, though in retrospect that makes total sense. Maybe we should through a "hint: @media" in the doc, hah ;)

I'm not certain if there's any technical problem with the pattern, the spaces are probably removed by CSS minifiers. We simply like the spaces for code style, and you know, automate-all-the-things (especially, the OCD things).

**Comment by jeddy3** [comment]
> I could take a swing at it sure.

Fantastic! It's best to wait for everyone else to weigh-in on the proposal before getting stuck in though.

> I've listened to talks on parsers l (Babel/ESLint), but I tend to get lost after they start talking about tree manipulations.

Ha, tell me about it! Don't worry though as PostCSS abstracts away all the complicated parser stuff and you'll just make use of the intuitive [PostCSS API](https://github.com/postcss/postcss/blob/master/docs/api.md). We've also written [a guide to getting started with rules](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#working-on-rules) to help people contribute.

The most similar existing rule is probably [`declaration-colon-space-after`](https://github.com/stylelint/stylelint/tree/master/src/rules/declaration-colon-space-after). You'll probably want to take ideas for the [docs](https://github.com/stylelint/stylelint/blob/master/src/rules/declaration-colon-space-after/README.md), [options validator](https://github.com/stylelint/stylelint/blob/master/src/rules/declaration-colon-space-after/index.js#L21-L29), [messages](https://github.com/stylelint/stylelint/blob/master/src/rules/declaration-colon-space-after/index.js#L12-L16) and [tests](https://github.com/stylelint/stylelint/blob/master/src/rules/declaration-colon-space-after/__tests__/index.js) from there.

You'll probably want to ignore the [whitespace checker stuff](https://github.com/stylelint/stylelint/blob/master/src/rules/declaration-colon-space-after/index.js#L31-L71) though (and it's not relevant here), and instead use a pattern similar to [this one](https://github.com/stylelint/stylelint/blob/master/src/rules/function-parentheses-space-inside/index.js#L73-L101) (where you'll be checking the value of `raw.afterName`).

Finally, [`at-rule-semicolon-newline-after`](https://github.com/stylelint/stylelint/blob/master/src/rules/at-rule-semicolon-newline-after/index.js#L26) will give you an idea of how to use the PostCSS API to *walk* through all the at-rules in a stylelint (spoiler: it's really simple!).

Hopefully everyone else agrees on the rule proposal, or I just wasted a bit of time pointing you in the wrong direction! :) Fingers crossed.

**Comment by davidtheclark** [comment]
The `at-rule-name-*` rules sound good to me.

**Comment by jeddy3** [comment]
Excellent.

@robcolburn Feel free to give this a shot. Please don't hesitate to ask questions here, or push up a WIP pull request if you've any queries :)

**Comment by davidtheclark** [comment]
Both rules have been added to master.

**Comment by robcolburn** [comment]
Thanks! These are great!
